packages/react/src/JsonFormsContext.tsx

```diff
@@ -126,11 +129,53 @@ const useEffectAfterFirstRender = (
   }, dependencies);
 };
 
+interface JsonFormsCoreStateProps
+  extends Omit<
+    JsonFormsInitStateProps,
+    | 'ajv'
+    | 'data'
+    | 'schema'
+    | 'uischema'
+    | 'validationMode'
+    | 'additionalErrors'
+  > {
+  core: JsonFormsCore;
+  config: Config;
+}
+interface JsonFormsStateProviderProps {
+  children: React.ReactNode;
+  initState: JsonFormsCoreStateProps;
+}
+
 export const JsonFormsStateProvider = ({
   children,
   initState,
   onChange,
-}: any) => {
+}: JsonFormsStateProviderProps & JsonFormsReactProps) => {
+  const onChangeRef = useRef(onChange);
+  useEffect(() => {
+    onChangeRef.current = onChange;
+  }, [onChange]);
+
+  const StateProvider =
+    initState.config.mode === 'controlled'
+      ? JsonFormsControlledStateProvider
+      : initState.config.mode === 'uncontrolled'
+      ? JsonFormsUncontrolledStateProvider
+      : JsonFormsMixedStateProvider;
```

Comment:
We should probably put the `StateProvider` in `useState` and issue a warning in case a developer tries to switch the mode after the fact.